parley/src/layout/cursor.rs

```diff
@@ -183,4 +233,250 @@ impl CursorPath {
     pub fn cluster<'a, B: Brush>(&self, layout: &'a Layout<B>) -> Option<Cluster<'a, B>> {
         self.run(layout)?.get(self.cluster_index)
     }
+
+    pub fn visual_line<'a, B: Brush>(&self, layout: &'a Layout<B>) -> Option<Line<'a, B>> {
+        layout.get(self.visual_line_index)
+    }
+}
+
+/// Returns a point that is falls within the vertical bounds of the given line.
+fn line_y<B: Brush>(line: &Line<B>) -> f32 {
+    line.metrics().baseline - line.metrics().ascent * 0.5
+}
+
+#[derive(Copy, Clone, PartialEq, Default, Debug)]
+pub struct Selection {
+    anchor: Cursor,
+    focus: Cursor,
+    h_pos: Option<f32>,
+}
+
+impl From<Cursor> for Selection {
+    fn from(value: Cursor) -> Self {
+        Self {
+            anchor: value,
+            focus: value,
+            h_pos: None,
+        }
+    }
+}
+
+impl Selection {
+    pub fn from_point<B: Brush>(layout: &Layout<B>, x: f32, y: f32) -> Self {
+        Cursor::from_point(layout, x, y).into()
+    }
+
+    pub fn from_byte_index<B: Brush>(layout: &Layout<B>, index: usize) -> Self {
```

Comment:
This should take an affinity parameter or similar.

Reply:
Done in as yet I pushed code